### PR TITLE
[CDEC-377] Reunite SafeCopy instances for core data types

### PR DESCRIPTION
**Pull request by erikd**
## Description

As per title

## Linked issue

https://iohk.myjetbrains.com/youtrack/issue/CDEC-377

## Type of change

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [x] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [ ] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [ ] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [ ] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [ ] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.

## Testing checklist

- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: core/src/Pos/Core/Txp/Tx.hs [hunk 1]
```diff
@@ -20,14 +20,16 @@ import           Universum
 
 import           Control.Lens (makeLenses, makePrisms)
 import           Control.Monad.Except (MonadError (throwError))
-import qualified Data.ByteString.Lazy as BSL
+
```
Comment (ruhatch) [inline core/src/Pos/Core/Txp/Tx.hs]: Whitespace wibble!

File: core/src/Pos/Core/Slotting/LocalSlotIndex.hs [hunk 2]
```diff
@@ -33,6 +34,7 @@ newtype LocalSlotIndex = UnsafeLocalSlotIndex
     { getSlotIndex :: Word16
     } deriving (Show, Eq, Ord, Ix, Generic, Hashable, Buildable, Typeable, NFData)
 
+
```
Comment (ruhatch) [inline core/src/Pos/Core/Slotting/LocalSlotIndex.hs]: Whitespace wibble!

File: core/src/Pos/Core/Txp/TxPayload.hs [hunk 3]
```diff
@@ -11,6 +11,8 @@ import           Universum
 import           Control.Lens (makeLenses)
 import           Control.Monad.Except (MonadError)
```
Comment (ruhatch) [inline core/src/Pos/Core/Txp/TxPayload.hs]: Another wibble!

File: acid-state-exts/src/Serokell/AcidState/Instances.hs [hunk 4]
```diff
@@ -25,9 +23,6 @@ instance (Eq a, Hashable a, SafeCopy a) => SafeCopy (HashSet a) where
     putCopy = contain . safePut . HS.toList
     getCopy = contain $ HS.fromList <$> safeGet
 
-instance (Eq a, Hashable a, SafeCopy a, SafeCopy b) => SafeCopy (HashMap a b) where
-    putCopy = contain . safePut . HM.toList
-    getCopy = contain $ HM.fromList <$> safeGet
```
Comment (parsonsmatt) [inline acid-state-exts/src/Serokell/AcidState/Instances.hs]: I'm actually not sure that  moving this "downstream" into `core` is the right idea. For things that are defined outside of our "universe", keeping them as upstream as possible makes sense to me. What do you think?

Comment (parsonsmatt) [reply acid-state-exts/src/Serokell/AcidState/Instances.hs]: That makes sense, thanks for clarifying :)

File: lib/src/Pos/SafeCopy.hs [hunk 5]
```diff
@@ -1,314 +0,0 @@
...
--- Plutus
-----------------------------------------------------------------------------
-
-instance Bi PLCore.Term => SafeCopy PLCore.Term where
-    getCopy = getCopyBi
-    putCopy = putCopyBi
-
-instance Bi PLCore.Program => SafeCopy PLCore.Program where
-    getCopy = getCopyBi
-    putCopy = putCopyBi
```
Comment (mhuesch) [inline lib/src/Pos/SafeCopy.hs]: Good riddance! ✌️

File: core/src/Pos/Core/Common/AddrSpendingData.hs [hunk 6]
```diff
@@ -123,3 +124,10 @@ addrSpendingDataToType =
         ScriptASD {} -> ATScript
         RedeemASD {} -> ATRedeem
         UnknownASD tag _ -> ATUnknown tag
+
+
+-- Define these at the end of the file to avoid TH staging issues.
+deriveSafeCopySimple 0 'base ''AddrSpendingData
+deriveSafeCopySimple 0 'base ''AddrType -- ☃
```
Comment (mhuesch) [inline core/src/Pos/Core/Common/AddrSpendingData.hs]: Not an actual issue, but do we know what ☃ snowman means?

File: core/src/Pos/Core/Update/SoftforkRule.hs [hunk 7]
```diff
@@ -58,3 +59,5 @@ deriveSimpleBi ''SoftforkRule [
         Field [| srMinThd       :: CoinPortion |],
         Field [| srThdDecrement :: CoinPortion |]
     ]]
+
+deriveSafeCopySimple 0 'base ''SoftforkRule -- 💋
```
Comment (mhuesch) [inline core/src/Pos/Core/Update/SoftforkRule.hs]: Wat 💋

File: crypto/Pos/Crypto/SafeCopy.hs [hunk 8]
```diff
@@ -1,89 +0,0 @@
...
-
-instance SafeCopy (ProxyCert w) where
-    putCopy (ProxyCert sig) = contain $ safePut sig
-    getCopy = contain $ ProxyCert <$> safeGet
-
-instance SafeCopy w => SafeCopy (ProxySignature w a) where
-    putCopy ProxySignature{..} = contain $ do
-        safePut psigPsk
-        safePut psigSig
-    getCopy = contain $ ProxySignature <$> safeGet <*> safeGet
```
Comment (mhuesch) [inline crypto/Pos/Crypto/SafeCopy.hs]: Another good riddance! 👋

File: core/src/Pos/Core/Common/AddrSpendingData.hs (hunk 6 above)
Comment (erikd) [reply core/src/Pos/Core/Common/AddrSpendingData.hs]: Nope, Just left it there for kicks :)

**Pull request merged**